src/stream/src/executor/aggregation/minput.rs

```diff
@@ -43,61 +74,1182 @@ impl<S: StateStore> MaterializedInputState<S> {
         columns: &[&ArrayImpl],
-        state_table: &mut StateTable<S>,
     ) -> StreamExecutorResult<()> {
-        self.inner
-            .apply_chunk(ops, visibility, columns, state_table)
-            .await
+        self.cache.apply_batch(StateCacheInputBatch::new(
+            ops,
+            visibility,
+            columns,
+            &self.cache_key_serializer,
+            &self.arg_col_indices,
+            &self.order_col_indices,
+        ));
+        Ok(())
     }
 
     /// Get the output of the state.
     pub async fn get_output(&mut self, state_table: &StateTable<S>) -> StreamExecutorResult<Datum> {
-        self.inner.get_output(state_table).await
+        if !self.cache.is_synced() {
+            let all_data_iter = iter_state_table(state_table, self.group_key.as_ref()).await?;
+            pin_mut!(all_data_iter);
+
+            let mut cache_filler = self.cache.begin_syncing();
+            #[for_await]
+            for state_row in all_data_iter.take(cache_filler.capacity()) {
+                let state_row = state_row?;
+                let cache_key = {
+                    let mut cache_key = Vec::new();
+                    self.cache_key_serializer.serialize_datums(
+                        self.state_table_order_col_indices
+                            .iter()
+                            .map(|col_idx| &(state_row.0)[*col_idx]),
+                        &mut cache_key,
+                    );
+                    cache_key
+                };
+                let cache_value = self
+                    .state_table_arg_col_indices
+                    .iter()
+                    .map(|i| state_row[*i].as_ref().map(ScalarImpl::as_scalar_ref_impl))
+                    .collect();
+                cache_filler.insert(cache_key, cache_value);
+            }
+        }
+        debug_assert!(self.cache.is_synced());
+        Ok(self.cache.get_output())
+    }
+}
+
+pub struct StateCacheInputBatch<'a> {
+    idx: usize,
+    ops: Ops<'a>,
+    visibility: Option<&'a Bitmap>,
+    columns: &'a [&'a ArrayImpl],
+    cache_key_serializer: &'a OrderedRowSerde,
+    arg_col_indices: &'a [usize],
+    order_col_indices: &'a [usize],
+}
+
+impl<'a> StateCacheInputBatch<'a> {
+    fn new(
+        ops: Ops<'a>,
+        visibility: Option<&'a Bitmap>,
+        columns: &'a [&'a ArrayImpl],
+        cache_key_serializer: &'a OrderedRowSerde,
+        arg_col_indices: &'a [usize],
+        order_col_indices: &'a [usize],
+    ) -> Self {
+        let first_idx = visibility.map_or(0, |v| v.next_set_bit(0).unwrap_or(ops.len()));
+        Self {
+            idx: first_idx,
+            ops,
+            visibility,
+            columns,
+            cache_key_serializer,
+            arg_col_indices,
+            order_col_indices,
+        }
+    }
+}
+
+impl<'a> Iterator for StateCacheInputBatch<'a> {
+    type Item = (Op, CacheKey, SmallVec<[DatumRef<'a>; 2]>);
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This logic seems very similar to the `[Data/Stream]ChunkIter`. Will a new struct of `ChunkRef` help?

```rust
struct DataChunkRef<'a> {
    columns: &'a [Column],
    vis: &'a Vis,
}

struct StreamChunkRef<'a> {
    data: DataChunkRef<'a>,
    ops: Ops<'a>,
}
```

https://github.com/risingwavelabs/risingwave/blob/06b079bfb76d1fca6ecae1f59754936673008aea/src/common/src/array/data_chunk_iter.rs#L48-L75

Reply:
Sure. I just want to state that there'll be more opportunities and less effort to polish these infrastructures if we prefer extracting the similar pattern to a single place (only if zero-cost). 🥰